lib/ddtrace/sampler.rb

```diff
@@ -92,38 +121,75 @@ def key_for(span)
   class PrioritySampler
     extend Forwardable
 
+    SAMPLE_RATE_METRIC_KEY = '_sample_rate'.freeze
+
     def initialize(opts = {})
-      @base_sampler = opts[:base_sampler] || RateSampler.new
-      @post_sampler = opts[:post_sampler] || RateByServiceSampler.new
+      @pre_sampler = opts[:base_sampler] || RateSampler.new
+      @priority_sampler = opts[:post_sampler] || RateByServiceSampler.new
     end
 
-    def sample(span)
-      return perform_sampling(span) unless span.context
-      return sampled_by_upstream(span) if span.context.sampling_priority
+    def sample?(span)
+      @pre_sampler.sample?(span)
+    end
 
-      perform_sampling(span).tap do |sampled|
-        span.context.sampling_priority = if sampled
-                                           Datadog::Ext::Priority::AUTO_KEEP
-                                         else
-                                           Datadog::Ext::Priority::AUTO_REJECT
-                                         end
+    def sample!(span)
+      # If pre-sampling is configured, do it first. (By default, this will sample at 100%.)
+      # NOTE: Pre-sampling at rates < 100% may result in partial traces; not recommended.
+      span.sampled = pre_sample?(span) ? @pre_sampler.sample!(span) : true
```

Comment:
We could move this logic into `#sample?` and just do `span.sampled = sample?` here

Reply:
No. In Ruby world, `!` means the function causes a mutation, and `?` returns `true/false` without mutation. If you put `sample!` inside `sample?` you could cause unintended side effects (e.g. adding a tag when you don't want to.)

Reply:
ah, yeah, I figured the `!` part, but I guess I read over it's usage here, :+1: